samples/Greengrass/src/main/java/greengrass/BasicDiscovery.java

```diff
@@ -0,0 +1,284 @@
+            final DiscoveryClientConfig discoveryClientConfig =
+                    new DiscoveryClientConfig(clientBootstrap, tlsCtx,
+                            new SocketOptions(), region, 1, proxyOptions);
+            final DiscoveryClient discoveryClient = new DiscoveryClient(discoveryClientConfig);
+            connection = getClientFromDiscovery(discoveryClient, clientBootstrap);
+
+            if(!connection.connect().get()) {
+                throw new RuntimeException("Failed to connect to GG Core using MQTT");
+            }
+
+            if ("subscribe".equals(mode) || "both".equals(mode)) {
+                final CompletableFuture<Integer> subFuture = connection.subscribe(topic, QualityOfService.AT_LEAST_ONCE, message -> {
+                    System.out.println(String.format("Message received on topic %s: %s",
+                            message.getTopic(), new String(message.getPayload(), StandardCharsets.UTF_8)));
+                });
+                System.out.println("Subscribe future returned: " + subFuture.get());
+            }
+
+            final Scanner scanner = new Scanner(System.in);
+            while(true) {
+                String input = null;
+                if("publish".equals(mode) || "both".equals(mode)) {
+                    System.out.println("Enter the message you want to publish to topic %s and press Enter. " +
+                            "Type 'exit' or 'quit' to exit this program.");
+                    input = scanner.nextLine();
+                }
+
+                if("exit".equals(input) || "quit".equals(input)) {
+                    System.out.println("Terminating...");
+                    break;
+                }
+
+                if("publish".equals(mode) || "both".equals(mode)) {
+                    System.out.println("Publishing message!");
+                    final CompletableFuture<Integer> publishResult = connection.publish(new MqttMessage(topic,
+                            input.getBytes(StandardCharsets.UTF_8)), QualityOfService.AT_LEAST_ONCE, false);
+                    Integer result = publishResult.get();
+                    System.out.println("Publish result code: " + result);
+                }
+            }
+        } catch (CrtRuntimeException | InterruptedException | ExecutionException ex) {
+            System.out.println("Exception encountered: " + ex.toString());
+            ex.printStackTrace();
+        }
+        finally {
+            if(connection != null && !connection.isNull()) {
+                CompletableFuture<Void> disconnected = connection.disconnect();
+                try {
+                    disconnected.get();
+                } catch (ExecutionException | InterruptedException e) {
+                    throw new RuntimeException("Problem disconnecting the MQTT connection cleanly: " + e.getMessage());
+                }
+            }
+        }
+        System.out.println("Complete!");
+    }
```

Comment:
Similar to Shadow or PubSubStress, add a CrtResource.waitForNoResources(); at the end here and then run the sample using CRT debugging (-Daws.crt.debugnative) and to track down any leaks you'll probably want to enable logging using system properties (https://github.com/awslabs/aws-crt-java/blob/master/src/main/java/software/amazon/awssdk/crt/Log.java#L32)